### ISSUE TITLE
API: Some entries not decrypted on fetchall

### ISSUE THREAD

**Issue by eglia** [fcturner/passwords#298]
I noticed a strange behaviour with my installation: two password entries are not decrypted when I fetch all passwords via API (`index.php/apps/passwords/api/0.1/passwords/`).
If i fetch them explicitly by their id however (`index.php/apps/passwords/api/0.1/passwords/123`), they are decrypted correctly.  Also in the web app they are displayed correctly.

I'm using nextcloud 10

Unfortunately, I can't provide more information at the moment. I didn't notice anything that would set the misbehaving entries apart and didn't see any interesting log entries. I will keep an eye out.

**Comment by fcturner** [comment]
How strange! The `findAll` query is exactly the same as the `find` query. Are those passwords shared, or were they shared and did you stop that?

**Comment by fcturner** [comment]
Another user in #294 said he had forgotten about the passwordsalt in config.pho when using NextCloud. Could that be it for you too?

**Comment by eglia** [comment]
I think I figured it out. The problem is in passwordservice.php, specifically the code you added to remove the share users from the list:
```
if ($arr[$row]['id'] == '0' && $api) {
  array_splice($arr, $row, 1);
}
```
You remove elements from the array while iterating over it. This leads to wrong array addressing and for each user entry that is removed, one password entry is skipped and at the end errors are produced because the array is no longer as long as expected.

**Comment by fcturner** [comment]
I see. But doesn't the PHP function array_splice shortens the array too? Then the array should be shorter afterwards. I thought I got that from the php.net website. Or is your app expecting a certain size? Could that be changed?

**Comment by eglia** [comment]
Yes, array_splice shortens the array which is exactly the problem:
Let's say the array has 10 elements at the start with one user entry at index 3. The loop iterates from index 0-9 (this is determined at the start of the loop). When it reaches index 3, it deletes the entry. The array now has indices 0-8, with new index 3 being old index 4 and so forth. The next iteration runs on new index 4, which was old index 5. Old index 4 was never processed. At the end of the loop it tries to access index 9, which no longer exists.
A solution to this problem could be to not directly delete the elements, but only remember the indices and then delete them after the loop. Or you could somehow modify the for loop indexing to compensate for the changed array size.
